File: DSA_Practice/LinkedListQuest/DoublyLL.java
import java.util.Scanner;

public class DoublyLL{
  static class Node{
        int data; 
        Node next, prev;

        Node(int data){
            this.data= data;
            this.next =null;
            this.prev = null;
        }
    }
    Node head = null, tail = null;
 
  
    public void insertAtBegin(int data){
        
           Node newNode =  new Node(data);
            newNode.next = head;
            newNode.prev= null;

            if(head !=null){
                tail = head;
                head.prev = newNode;             
            }
            head = newNode;
        }

    public void printList(){
        Node temp = head;
        while(temp!= null){
            System.out.print(temp.data+"->");
            temp = temp.next;
        }
        System.out.print("->END");
        System.out.println();
    }

    public void insertEnd(int val){
        Node newNode = new Node(val);
        if(head == null){
            insertAtBegin(val);
            return;
        }
        Node temp = head;
        while(temp.next !=null){
            temp = temp.next;
           
        }
         newNode.prev =temp;
         temp.next = newNode;
    }

    public void insertMiddle(int val, int pos){
        int count=1;
        Node newNode = new Node(val);
        if(head == null){
            insertAtBegin(val);
            return;
        }
        Node temp = head;
        while(count != pos){
            temp = temp.next;
            count++;
        }
        newNode.prev = temp.prev;
        newNode.next= temp;
        temp.prev.next = newNode; 
    }

    public static void main(String[] args) {
        
        DoublyLL dl = new DoublyLL();
        dl.insertAtBegin(5);
        dl.insertAtBegin(7);
        dl.insertAtBegin(9);
        dl.insertAtBegin(11);
        dl.insertAtBegin(13);
        dl.insertAtBegin(15);
        dl.insertAtBegin(17);
        dl.printList();
      
       dl.insertEnd(3);
       dl.insertEnd(1);
       dl.insertEnd(0);
       dl.printList();

       dl.insertMiddle(14, 3);
       dl.printList(); 
       dl.insertMiddle(8, 7);
       dl.printList();
       
       

    }
}

File: DSA_Practice/CppDemooo/DoublyLL.cpp
#include<bits/stdc++.h>
using namespace std;

class Node{
    public:
     int data;
     Node *next, *prev;

     Node(int value){
        data = value;
        prev = next = nullptr;
     }
};

class DoublyLL{
private:
    Node *head;
    int totalNodes;

public:
    DoublyLL(){ 
        head =  nullptr;
        totalNodes = 0;
    }

    void insertAtBegin(int );
    void insertAtPosition(int , int);
    void insertAtEnd(int);

    void countNodes();
    void displayDLL();

    void deleteAtBegin();
    void deleteAtPosition(int);
    void deleteAtEnd();

};


  void DoublyLL :: insertAtBegin(int value){
        Node *newNode = new Node(value);
    // Base case:
        if( head == nullptr){
            head = newNode;
        }
        else{
            newNode->next = head;
            head->prev = newNode;
            head = newNode;
        }
        totalNodes++;
    }


    void DoublyLL :: insertAtPosition(int value, int position){
    // Base case:
     if (position < 1 || position > totalNodes + 1) {
        cout << "Invalid position!" << endl;
        return;
    }
     Node *newNode = new Node(value);
     if(position == 1){
        insertAtBegin(value);
        return ;
     }

     // Insert at the end
     if (position == totalNodes+1){
            insertAtEnd(value);
            return ;
     }
    // Insert At Specific Position:

        Node* temp = head;
        for (int i = 1; i < position - 1; i++) {
                   temp = temp->next;
        }
        newNode->next = temp->next;
        newNode->prev = temp;
        temp->next->prev = newNode;
        temp->next = newNode;
        totalNodes++;

    }

    void DoublyLL :: insertAtEnd(int value){
        Node *newNode = new Node(value);
    // Base case:
        if( head == nullptr){
            head = newNode;
            totalNodes++;  
            return;
        }
    Node *temp = head;
        while(temp->next != nullptr){
            temp= temp->next;
        }
         temp->next = newNode;
         newNode->prev = temp;
         totalNodes++;   // increment the node after inserting at end of the node.
   }


void DoublyLL :: deleteAtBegin(){
    Node *temp = head;
    
    if (head == nullptr) {
        cout << "List is empty. Cannot delete." << endl;
        return;
    }
    // If there's only one node
    if (head->next == nullptr) {
        head = nullptr;
    } 
    else {
            head = head->next;
            head->prev = nullptr;
            return ;
    }
            delete temp;
            totalNodes--;
}


void DoublyLL :: deleteAtPosition(int position){
     Node *temp = head;
    
    if (head == nullptr) {
        cout << "List is empty. Cannot delete." << endl;
        return;
    }

    // when position ==1 is head itself
    if(position == 1){
        head = head->next;
        if (head != nullptr) {
            head->prev = nullptr; 
        }
        delete temp;
        totalNodes--;
        return;
    }

    for(int i=1; i<position; i++){
        temp = temp->next;
    }
    // Updating pointer here
    if (temp->next != nullptr) {
        temp->next->prev = temp->prev;
    }
    temp->prev->next = temp->next;

    delete temp;
    totalNodes--;
}



void DoublyLL :: deleteAtEnd(){ 
    Node *temp = head;
    
    if (head == nullptr) {
        cout << "List is empty. Cannot delete." << endl;
        return;
    }
    // If there is only-one node in the list
    if(head->next == nullptr){
        head = nullptr;
        delete temp;
        totalNodes--;
        return;
    }

    while(temp->next !=nullptr){
        temp = temp->next;
    }
        temp->prev->next = nullptr;
        delete temp;

        totalNodes--;    
}

 void DoublyLL :: countNodes(){
        cout<<"Total number of nodes in the list are:"<<" "<<totalNodes <<endl; 

 }

 void DoublyLL :: displayDLL(){
    if(head == nullptr){
        cout<<"Node list is empty:" <<endl;
    }
    else{
        Node *temp = head;
        while(temp !=nullptr){
            cout<<temp->data <<" ";
            temp = temp->next;
        }
        cout<<endl;
    }
 }
int main(){
    DoublyLL doublyll;
// Insert at Begin:
    doublyll.insertAtBegin(15);
    doublyll.insertAtBegin(13);
    doublyll.insertAtBegin(11);
    doublyll.insertAtBegin(9);
    doublyll.insertAtBegin(7);
doublyll.displayDLL();
    doublyll.countNodes();

// Insert at End:
    doublyll.insertAtEnd(17);
    doublyll.insertAtEnd(19);
    doublyll.insertAtEnd(21);
doublyll.displayDLL();
    doublyll.countNodes();


// Insert at Specifivc Position:
    doublyll.insertAtPosition(12,4);
doublyll.displayDLL();
    doublyll.countNodes();



// Delete Operations:

// Delete at Begin:
    doublyll.deleteAtBegin();
doublyll.displayDLL();

// Delete at Specific Position:
    doublyll.deleteAtPosition();
doublyll.displayDLL();


// Delete at End:
    doublyll.deleteAtEnd();
doublyll.displayDLL();

    return 0;
}
